HW3/P5/label_regions.cl

```diff
@@ -80,20 +80,57 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
-    
+    // Overwritten code for PART 2
+    // if (old_label < w * h) {
+    //     // Grab grandparent
+    //     buffer[buf_y * buf_w + buf_x] = labels[old_label];
+    // }
+
+    // PART 4
+    // Update workgroup labels using single thread
+    if (lx == 0 && ly == 0) {
+        // Keeps track of previous key and result
+        int prev_label = -1;
+        int prev_result;
+
+        // Iterate through entire buffer
+        for (int i = 0; i < buf_w * buf_h; i++) {
+            int temp_label = buffer[i];
+
+            if (temp_label < w * h) {
+                // Reset if previous is not the same as current label
+                if (prev_label != temp_label) {
+                    prev_label = temp_label;
+                    prev_result = labels[prev_label];
+                }
+                buffer[i] = prev_result;
+            }
+        }
+    }
+
     // stay in bounds
-    if ((x < w) && (y < h)) {
+    if ((x < w) && (y < h) && (old_label < w * h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
+        
+        // Update new label
         new_label = old_label;
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.